### DISCUSSION TITLE
Midnight Club: LA (Some overlap with RDR)

### DISCUSSION THREAD

**Discussion by bomabomaboma** [xenia-canary/game-patches#14] [General]
FPS-Heavy Shader Draw Commands:
AmbientOcclusion_draw
AmbientOcclusion_drawskinned
shadowBlend_draw
shadowDepth_draw
shadowAlphaDepth_draw

Shader Commands To Disable In CPU-Limited Situations: (Usually z-prepass is a good thing, this might actually worsen framerates. Still worth investigating.)
zprepassAlpha_draw
zprepass_draw
zprepassCrossFade_draw

Other Shader GPU-Heavy Commands:
generate_impostor_draw
generate_impostor_drawskinned
generate_night_impostor_drawskinned
generate_shadow_impostor_draw 		 (These are mostly for tree and foliage shadows)
generate_shadow_impostor_drawskinned (These are mostly for tree and foliage shadows)

Commands That Call Object-Specific Parameters Within The Hundreds upon Hundreds of Configuration Files: (This could be very useful.)
CubeShadowMapSampler
CubeShadowMap
ShadowCollectorSampler
ShadowCollectorTex

Value "rage__CShadowMapTile", are the shadow maps for every mesh that casts real-time shadows. "bamblight" or "amblight" is their SSAO, if applicable. Sending those shadow maps, with their included depth maps, bias, camera off-set, EVERYTHING, back to the shader as a "0" disables it all.

**Comment by bomabomaboma** [comment]
Default
*Note the frametime graph at the second half of the graph.
![default](https://user-images.githubusercontent.com/81534189/115977173-6ebcab80-a52a-11eb-850c-07fb1169428d.jpg)

No-Shadows (SSAO still enabled)
![noshadows](https://user-images.githubusercontent.com/81534189/115977186-8f850100-a52a-11eb-91e8-577434af8bc8.jpg)

Can't stress this enough, this game has a bad, bad bandwidth problem doing texture streaming and shadow map calling/drawing.
This turns an unplayable mess on my system into a smooth, fun experience.

**Comment by illusionyy** [comment]
Oh damn, totally missed this one.

**Reply by illusionyy** [reply]
Btw I tried to patch shadow stuff for rdr, not much luck since it loads them from file later populates it in some way, not to mention it does it all from a single instruction.. it may be sth or lhz sure but I can't seem to separate it.

**Reply by illusionyy** [reply]
Perhaps loading memory patterns and do some sort of memscan and offsetting it may work. But since there's many variables it will be very long. Will look at it later. Gonna poke around with MCLA now.

**Reply by bomabomaboma** [reply]
> Btw I tried to patch shadow stuff for rdr, not much luck since it loads them from file later populates it in some way, not to mention it does it all from a single instruction.. it may be sth or lhz sure but I can't seem to separate it.

You can't force specific bucket item sizes, or interrupt return values from querying the bucket sizes? Bucket size adjustment will leave in real-time shadowing from the skybox on items with an "imposter" value, which are mostly trees and some buildings, but other dynamic shadows are disabled. Disabling shadows that way provide a larger performance gain than zeroing out shadowmap tile sizes. But, there are a lot of buckets that need adjusting...It wouldn't be a short patch.

I'm not surprised, the world is constantly streaming in new parameters from separate branches of files based on the area of the map the player is currently located.

MC:LA is more straightforward.
Again, as is the case with RDR, opening the shader .fxc files in hex editor will show you the string names of the scripting processes that perform collection of shadow and lighting parameters of all objects to be sent to the shader.

**Reply by illusionyy** [reply]
Hmm, have you found anything related to framerate timing? MCLA for some reason uses fixed rate which means everything slows down when fps isn't at 30. Strange choice.

**Reply by illusionyy** [reply]
Also a reason why the game feels sooo sluggish

**Reply by bomabomaboma** [reply]
tune\types\fragments\0x6f128fd7 has two values of interest
- GlobalMaxDrawingDistance (default value is 3000)
- BreakingFrameRateLimit (default value is 10)


Haven't tested BreakingFrameRateLimit yet. Don't know what that does. Maybe that's what's going on? When you drop below 20fps off target 30fps (10 differential) that cache and memory pools are dynamically changed, and that presents as slow-motion for some reason or another? I'll try changing that to 20 when I have some time to take screenshots and compare.

GlobalMaxDrawingDistance only applies to geometry mipmaps and textures resolution. For RTV, that could be useful. Certainly in RAGE games when loading new mipmap/LOD values when you're transitioning to a new part of the map can lead to horrific framepacing.

**Comment by Xbox360forever** [comment]
Hello, I'm very new of Github and sorry to write you here, I didn't know where otherwise. As reported in my youtube channel, I'm trying to apply the Xenia patches, expecially to unlock the framerate, on a real Xbox 360 modifying the .xex files. Some of them work well (Half Life 2, thanks to Margen67), Gears of War, Shadows of the damned, other not. I've also unlocked the framerate some games by myself, like Ace Combat 6, King Kong, Batman Arkham City (see the relative videos on my youtube channel for the hex addresses to be changed). Unfortunately, some of Xenia patches don't work or work but, I presume but not sure, I need also to turn off the vertical sync. One that doesn't work is Midnight Club: LA. Is there a way to report how to turn off the vsync in the patches you make? I'll be very glad. ps. I've only basic of C++, not of assembler, so my approach is to edit the hex code with tools like Peek Poker and HxD, changing some typical values I've notice are reported in the Xenia patches (eg. 39 60 00 02 48 00 00, changing 02 to 01). Thanks in advance.
